pkg/queryfrontend/queryinstant_codec.go

```diff
@@ -283,22 +293,66 @@ func vectorMerge(resps []*queryrange.PrometheusInstantQueryResponse) *queryrange
 	if len(output) == 0 {
 		return &queryrange.Vector{
 			Samples: make([]*queryrange.Sample, 0),
-		}
+		}, nil
 	}
 
-	keys := make([]string, 0, len(output))
-	for key := range output {
-		keys = append(keys, key)
+	type pair struct {
+		metric string
+		s      *queryrange.Sample
 	}
-	sort.Strings(keys)
 
+	samples := make([]*pair, 0, len(output))
+	for k, v := range output {
+		samples = append(samples, &pair{
+			metric: k,
+			s:      v,
+		})
+	}
+
+	sort.Slice(samples, func(i, j int) bool {
+		// Order is determined by the sortFn in the query.
+		if sortAsc {
+			return samples[i].s.Sample.Value < samples[j].s.Sample.Value
+		} else if sortDesc {
+			return samples[i].s.Sample.Value > samples[j].s.Sample.Value
+		} else {
+			// Fallback on sorting by labels.
+			return samples[i].metric < samples[j].metric
+		}
+	})
 	result := &queryrange.Vector{
 		Samples: make([]*queryrange.Sample, 0, len(output)),
 	}
-	for _, key := range keys {
-		result.Samples = append(result.Samples, output[key])
+	for _, p := range samples {
+		result.Samples = append(result.Samples, p.s)
 	}
-	return result
+	return result, nil
+}
+
+func parseQueryForSort(q string) (bool, bool, error) {
+	expr, err := promqlparser.ParseExpr(q)
+	if err != nil {
+		return false, false, err
+	}
+	var sortAsc bool = false
+	var sortDesc bool = false
+	done := errors.New("done")
+	promqlparser.Inspect(expr, func(n promqlparser.Node, _ []promqlparser.Node) error {
```

Comment:
I wonder if we should just check the root expression and decide how to sort. You example query is a binary expression and I think we have to sort the end result by labels, not by values. Worth double checking through.

Reply:
In the below cases Prometheus sorts the final result by value:
- `2 + sort (sum by (instance, code) (prometheus_http_requests_total))`
- `sort (sum by (instance, code) (prometheus_http_requests_total)) + sort _desc(sum by (instance, code) (prometheus_http_requests_total))`

But `topk` and `bottomk` will break the logic:
- `topk (3, sort(sum by (instance, code) (prometheus_http_requests_total)))` shouldn't be sorted in ascending order.
- `bottomk (3, sort_desc(sum by (instance, code) (prometheus_http_requests_total)))` shouldn't be sorted in descending order.
- Here, `topk` and `bottomk` decides the final sort order and the results should be sorted by value in descending order.

I think we need to re-sort the final results for `topk` and `bottomk` functions also in the query-frontend.
When walking the AST, the first of `topk`, `bottomk`, `sort`, `sort_desc` should decide the sort order.

Reply:
> You example query is a binary expression and I think we need to sort the end result by labels, not by values. Worth double checking through.

I think in current promql implementation it doesn't sort results by labels at all in this case. I also doubt if it really sort the final results by the first `sort` or `sort_desc`.

Reply:
Yeah I am not sure if this is part of the spec, or just a side effect of the internals of the engine. 

> When walking the AST, the first of topk, bottomk, sort, sort_desc should decide the sort order.

Hm but if those are wrapped by something like a `sum`, shouldn't the sum decide the ordering? My vote would be to keep things simple and sort based on the root expression. If the root is `sort` or `sort_desc`, we sort by values. `topk` and `bottomk` only sort by values for instant queries. For everything else I would vote to sort by labels, or not sort at all. Otherwise we might be adding complexity due to implementation details of how the prometheus engine works. Wdyt?

Reply:
@harry671003 is only changing instant query merging now.

Indeed I think you are correct to only check root expression. But I can see we will still mismatch the results from the promql engine for the binary expression example. Same for topk and bottomk.

What I am thinking about is only shard if the root expression is a func and it contains the sorting methods. If it is other expressions we just disable sharding to keep consistent results.

Reply:
Probably it is fine to be inconsistent with Promql engine as the behavior is kind of undefined in this case. 
I am okay if we only check root expression as well and sort by labels in other case.

Reply:
> @harry671003 is only changing instant query merging now.

Ah ok got it. This also makes sense since range queries seem to always be sorted by values. 

Looking at the examples again, it does indeed make sense to present a binary expression with a constant in sorted order, e.g`(2 + sort(...))`. For something like `sum(sort)` I would not expect the result to be sorted, but I guess there is no harm if it is. 

For operations like `topk by (label)`, the sorting is only maintained within a group, and not globally. So values from one group appear together in Prometheus. Do we want to take that into account as well?

Reply:
> For operations like topk by (label), the sorting is only maintained within a group, and not globally. So values from one group appear together in Prometheus

The current sort by series labels way produces incorrect order for `topk`. But it sounds very hard to maintain the order within the group in our final results after merge. Shall we just skip sharding in this case?

Reply:
Let me try to summarize what I understood from the discussion above:
- For `sort` and `sort_desc` we want to continue to shard the query and sort the results at the end.
- For `topk` and `bottomk` we want to disable sharding because sorting at the end will break `topk by (label)`.

Reply:
Looking at this again, for `topk` and `bottomk` we don't need actually do anything because the shards themselves will already be sorted. Since `topk/bottomk by` does not require a global sort, it should be safe to keep sharding these expressions. 

So I think we're good with this implementation :)

Reply:
> Looking at this again, for topk and bottomk we don't need actually do anything because the shards themselves will already be sorted.

Even if the shards are sorted, when we merge here in https://github.com/thanos-io/thanos/blob/main/pkg/queryfrontend/queryinstant_codec.go#L259 won't the order change?

Reply:
After more discussions with @yeya24, here's the final proposal we came up with that works mostly:
- If the root expr is `topk` or `bottomk`, the response will be concatenated without a global sort.
- If the AST contains `sort` or `sort_desc`, the response will be sorted by values.
- Else, the response will be sorted by labels.

Examples:
| # | Query                                                                | Merge mechanism                | Prometheus Link                                                                                                                                                                                               | Remarks                     |
|---|----------------------------------------------------------------------|--------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------|
| 1 | `1 + sort(sum by (code) (http_requests_total) )`                     | Merge and then sort by values  | [link](https://prometheus.demo.do.prometheus.io/graph?g0.expr=1%20%2B%20sort(sum%20by%20(code)%20(prometheus_http_requests_total)%20)&g0.tab=1&g0.stacked=0&g0.range_input=1h)                                | Prometheus sorts by value.  |
| 2 | `sort(topk by (code) (10, http_requests_total))`                     | Merge and then sort by values  | [link](https://prometheus.demo.do.prometheus.io/graph?g0.expr=sort(topk%20by%20(code)%20(10%2C%20prometheus_http_requests_total))&g0.tab=1&g0.stacked=0&g0.range_input=1h)                                    | Prometheus sorts by value.  |
| 3 | `topk by (code) (10,  http_requests_total)`                          | Merge responses but don't sort | [link](https://prometheus.demo.do.prometheus.io/graph?g0.expr=topk%20by%20(code)%20(10%2C%20%20http_requests_total)&g0.tab=1&g0.stacked=0&g0.range_input=1h)                                                  | Prometheus doesn't sort.    |
| 4 | `topk(5, http_requests_total) by (code) + sort(http_requests_total)` | Merge and then sort by values  | [link](https://prometheus.demo.do.prometheus.io/graph?g0.expr=topk(5%2C%20prometheus_http_requests_total)%20by%20(code)%20%2B%20sort(prometheus_http_requests_total)&g0.tab=1&g0.stacked=0&g0.range_input=1h) | Prometheus doesn't sort.    |
| 5 | `sort(http_requests_total) + topk(5, http_requests_total) by (code)` | Merge and then sort by values  | [link](https://prometheus.demo.do.prometheus.io/graph?g0.expr=sort(prometheus_http_requests_total)%20%2B%20topk(5%2C%20prometheus_http_requests_total)%20by%20(code)&g0.tab=1&g0.stacked=0&g0.range_input=1h) | Prometheus sorts by value.  |


There is discrepancy for case 4 where Prometheus doesn't sort the responses but Thanos will sort by value.

I believe the above approach will give reasonable compatibility with Prometheus without increasing the complexity in the merging logic in the frontend.

@fpetkovski @yeya24 - Please let me know if this approach makes sense.

Reply:
Thanks for the breakdown, this is what we needed and it makes sense to me 👍 

In the AST analysis for case 2. we can also look at the root expression, and if it's a binary one we can look only at the LHS and RHS but not any deeper than that. Not a blocker though, either approach would works close enough.